client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -354,70 +362,47 @@ public void processingFailure(Exception error) {
             try {
                 connection.send(new Append(segmentName, writerId, eventNumber, Unpooled.wrappedBuffer(event.getData()),
                                            event.getExpectedOffset()));
             } catch (ConnectionFailedException e) {
                 log.warn("Connection " + writerId + " failed due to: ", e);
-                getConnection(); // As the messages is inflight, this will perform the retransmition.
+                try {
+                    getConnection(); // As the messages is inflight, this will perform the retransmition.
+                } catch (SegmentSealedException e2) {
+                    return;
+                }
             }
         }
     }
     
     /**
-     * Blocking call to establish a connection and wait for it to be setup. (Retries built in)
+     * Establish a connection and wait for it to be setup. (Retries built in)
      */
-    ClientConnection getConnection() throws SegmentSealedException {
+    CompletableFuture<ClientConnection> getConnection() throws SegmentSealedException {
         checkState(!state.isClosed(), "LogOutputStream was already closed");
         if (state.isAlreadySealed()) {
             throw new SegmentSealedException(this.segmentName);
         }
-        return retrySchedule.retryingOn(ConnectionFailedException.class).throwingOn(SegmentSealedException.class).run(() -> {
-            setupConnection();
-            return state.waitForConnection();
-        });
-    }
-
-    @VisibleForTesting
-    void setupConnection() throws ConnectionFailedException {
-        if (state.getConnection() == null) {
-            synchronized (connectionEstablishmentLock) {
-                if (state.getConnection() == null && !state.isAlreadySealed()) {
-                    log.info("Fetching endpoint for segment {}", segmentName);
-                    CompletableFuture<ClientConnection> newConnection = controller.getEndpointForSegment(segmentName)
-                            .thenCompose((PravegaNodeUri uri) -> {
-                                log.info("Establishing connection to {} for {}", uri, segmentName);
-                                return connectionFactory.establishConnection(uri, responseProcessor);
-                            });
-                    ClientConnection connection = getAndHandleExceptions(newConnection, ConnectionFailedException::new);
-                    state.newConnection(connection);
-                    SetupAppend cmd = new SetupAppend(requestIdGenerator.get(), writerId, segmentName);
-                    try {
-                        connection.send(cmd);
-                    } catch (Exception e) {
-                        state.failConnection(e);
-                        throw e;
-                    }
-                }
-            }
-        }
+        CompletableFuture<ClientConnection> future =  new CompletableFuture<>();
+        state.setupConnection.runReleaserAndAwait(this::reconnect, future);
```

Comment:
This leads to a nested invocation of `runReleaserAndAwait` causing the  the infinite retry connect logic to be a no-op.